Add vulnerability remediation procedure (#8)
ros-security/community: vuln-remediation.md

# Vulnerability Remediation

> This document describes the typical process for remediating security vulnerabilies in ROS 2, including those submitted through the process  described in [REP 2006, ROS 2 Vulnerability Disclosure Policy](https://www.ros.org/reps/rep-2006.html).

Reviewer: ```suggestion
This document describes the typical process for remediating security vulnerabilies in ROS 2, including those submitted through the process described in [REP 2006, ROS 2 Vulnerability Disclosure Policy](https://www.ros.org/reps/rep-2006.html).
```

Author: Updated in 1143387b70d57380cf6501ac1090bf1c3f1e1411

### Coordinator

The individual responsible for facilitating the coordinated response process. This includes triaging the initial vulnerability report, identifying a Maintainer responsible for fixing the vulnerability, and tracking the report to a resolution. The Coordinator must, to the best of his or her ability, protect the confidentiality of the vulnerability until a fix has been published.

### Maintainer

The individual or organization responsible for updating the code base to eliminate the vulnerability. The Maintainer also must, to the best of his or her ability, protect the confidentiality of the vulnerability until a fix has been published.

These roles generally align with standard definitions in [The CERT Guide to Coordinated Vulnerability Disclosure](https://vuls.cert.org/confluence/display/CVD/3.+Roles+in+CVD); however, our *Maintainer* assumes the duties of *Deployer*, and the *Coordinator* also assumes the duties of the *Vendor* and works with Open Robotics as needed.

## Process

This process begins with a vulnerability report sent to [security@openrobotics.org](mailto:security@openrobotics.org).

> 1. **Assign a Coordinator.** Members of the `security` distribution list must reach consensus on the individual to take ownership of the issue. OpenRobotics will assign a Coordinator when consensus cannot be reached.

Reviewer: ```suggestion
1. **Assign a Coordinator.** Members of the `security` distribution list must reach consensus on the individual to take ownership of the issue. Open Robotics will assign a Coordinator when consensus cannot be reached.
```

Author: Updated in 1143387b70d57380cf6501ac1090bf1c3f1e1411

      > SUBJECT: [HIGH | Medium | Low] risk security vulnerability in [package name]
      > 
      > A [HIGH | Medium | Low] risk security vulnerability has been identified in [package name].  Please contact me at [contact] to discuss the vulnerability details.
      > 
      > [include if needed] Based on the severity of the vulnerability, the details are particularly sensitive and need to be protected through secure communications. Please send me your public PGP key through which you can receive secure email; if you are unable to send and receive secure email, please let me know and we will establish another secure channel.
      > 
      > [Attach a PGP key or a Key ID]
      >
      > [Sign the email with the PGP key]
      >

>    1. **Coordinator registers a CVE.** Seek help from the [ROS 2 security working group](https://github.com/ros-security/community#communication-channels) if necessary to reserve the CVE. The reserved CVE should not include any detailed information; this will be added after disclosure. All subsequent communications should include the CVE numnber for traceability.

Reviewer 2: ```suggestion
   1. **Coordinator registers a CVE.** Seek help from the [ROS 2 security working group](https://github.com/ros-security/community#communication-channels) if necessary to reserve the CVE. The reserved CVE should not include any detailed information; this will be added after disclosure. All subsequent communications should include the CVE number for traceability.
```

Author: Fixed in 44179b32ecadc54d74ed05f01ba81d9b3b58c2c9

1. **Remediation**

   1. **Maintainer fixes the vulnerability.** Work should be done locally as much as possible; when the fix is pushed to github the fix becomes visible to the public even though the patch has not been released to users. The final pull request for the fix should include a reference to the CVE in comments for traceability.

>    1. **Coordinator tracks to completion.**  This likely will mean periodic update requests to the Maintainer until the Maintainer publishes a fix. Seek assistance from the Security WG if needed.

Reviewer 2: Using the same wording for referring to the ROS 2 Security Working Group throughout the document would be beneficial

Author: Updated all references to the "Security Working Group" (capitalized) in 44179b32ecadc54d74ed05f01ba81d9b3b58c2c9

>    1. **Coordinator plans for vulnerability disclosure.** For high risk vulnerabilities or fixes with significant impact, create a communications plan for patch release and full disclosure. The plan should take input from the Maintainer and the Reporter.

Reviewer 2: Do "Coordinator"s have access/power to make ROS syncs happen ? this would be a prerequisite for making sure the fix makes it into the main repository in a timely matter after fix is published

Author: No, "Coordinator" is purely a communication/facilitation/administrative role. Any time the coordinator hits a roadblock or needs assistance, they should escalate, typically to either OpenRobotics or the TSC.

1. **Disclosure**

   1. **Maintainer conducts patch pre-notification.** When the fix is ready for release, the Maintainer, at his or her discretion, will notify select individuals that the fix is ready. This notification should be specifically to quickly remediate any build issues caused by the patch, to merge the pull request, monitor as the patch is bloomed, and finally confirm the fix is sync'd into the ROS distro.

   1. **Coordinator notifies the Reporter that the patch has been released.** At this point the patch is available as an update to ROS users.

   1. **Disclose the patch to a broader audience as desired.** Follow the vulnerabilty disclosure plan if one exists. Consider notifying the ROS community of the patch availability; for example, with a general post to [ROS discourse](https://discourse.ros.org/).

   1. **Coordinator updates CVE details.** Include a link to the patch and final CVSS details. 

## Additional Guidance

### Identifying the Maintainer

> If the Maintainer of the vulnerable package is not well known, review recent activity for the package. Contact pull request approver(s) to find a responsible person.

Reviewer: Another good way to find a maintainer is to look for the `<maintainer>` tag in the package.xml.

Author: Added this as a hint in 1143387b70d57380cf6501ac1090bf1c3f1e1411

> If the Maintainer of the vulnerable package is not well known, check for a <maintainer> tag in the package's `package.xml` file. Also review recent activity for the package, and contact a recent pull request approver(s) to find a responsible person.

Reviewer 2: ```suggestion
If the Maintainer of the vulnerable package is not well known, check for a <maintainer> tag in the package's `package.xml` file. Also review recent activity for the package, and contact a recent pull request merger(s) to find a responsible person.
```

People reviewing / approving Pull Requests may not be the same people as the ones merging and releasing code updates

Author: Updated in 44179b32ecadc54d74ed05f01ba81d9b3b58c2c9

### Requesting information from the Reporter

Reporters may have a wealth of additional details about the vulnerability which they are willing to share when asked.  Consider requesting some or all of the following:

 - Operating system
 - ROS distro

>  - ROS package

Reviewer 2: ```suggestion
 - ROS package and ROS package version
```

Author: Updated in 44179b32ecadc54d74ed05f01ba81d9b3b58c2c9

### The `security@openrobotics.org` distribution list

This email distribution list is to be used exclusively for the public to report ROS vulnerabilities. Unless explicitly labeled otherwise, conversations to the list are considered private and confidential. Follow the [CISA Traffic Light Protocol (TLP)](https://www.cisa.gov/tlp) guidance for TLP:RED information usage and sharing.

Distribution list membership will be maintained by Open Robotics. Membership must be broad enough to ensure coverage for timely response to Reporters. No email aliases or shared mailboxes may be included in the list.

Individuals added to the distribution list must adhere to the following principles:
 - You must be able to take ownership of reported vulnerabilities and act as the Coordinator through remediation. When high risk vulnerabilities are reported, this means vulnerability coordination becomes your highest priority until a fix is released.

>  - You must reach out to the Security Working Group or individual members of the Working group when you need assistance in handling vulnerability reports.

Reviewer 2: RE: use single wording for referring to the working group

We seem to be conflating the coordinators and the working group. Currently channels to contact the working group are public and meant to be public. How would one coordinator go through contacting the working group with vulnerability details? what is the responsibility of the working group in these cases ? (non-disclosure, obligation to follow through the remediation etc)

Author: Based on our WG discussions, this is written so the coordinator plays an operational role and the working group plays a purely advisory role. The coordinator may not need help with security-specific tasks, so we should be there to advise / support. That could include things like:

 - Whether a CVE is warranted
 - How to request a CVE number
 - Help with CVSS / risk scoring
 - Reaching out to a third party to help with vuln coordination

No special communications channel exists for the Security WG for vuln coordination, so this would come in on one of our [public channels](https://github.com/ros-security/community#communication-channels), but would fork to a private / encrypted conversation to discuss embargoed topics.

 - You must be able and willing to handle PGP-encrypted email.

> 

Reviewer 2: Why 2 extra blank lines ?

Author: Thanks, removed in 44179b32ecadc54d74ed05f01ba81d9b3b58c2c9
